File: src/java/src/androidTest/java/com/tencent/wcdb/winq/ColumnConstraintTest.java
package com.tencent.wcdb.winq;// Created by chenqiuwen on 2023/4/6.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import static com.tencent.wcdb.TestUtility.TestTool.winqEqual;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class ColumnConstraintTest {
    @Test
    public void test() {
        winqEqual(new ColumnConstraint().primaryKey(), "PRIMARY KEY");
        winqEqual(new ColumnConstraint("testColumnConstraint").primaryKey(),
                "CONSTRAINT testColumnConstraint PRIMARY KEY");
        winqEqual(new ColumnConstraint().primaryKey().order(Order.Asc), "PRIMARY KEY ASC");
        winqEqual(new ColumnConstraint().primaryKey().conflict(ConflictAction.Abort),
                "PRIMARY KEY ON CONFLICT ABORT");
        winqEqual(new ColumnConstraint().primaryKey().autoIncrement(), "PRIMARY KEY AUTOINCREMENT");
        winqEqual(new ColumnConstraint("testColumnConstraint").notNull(),
                "CONSTRAINT testColumnConstraint NOT NULL");
        winqEqual(new ColumnConstraint().notNull().conflict(ConflictAction.Abort),
                "NOT NULL ON CONFLICT ABORT");
        winqEqual(new ColumnConstraint("testColumnConstraint").unique(),
                "CONSTRAINT testColumnConstraint UNIQUE");
        winqEqual(new ColumnConstraint().defaultTo(1), "DEFAULT 1");
        winqEqual(new ColumnConstraint().defaultTo(false), "DEFAULT FALSE");
        winqEqual(new ColumnConstraint().defaultTo("abc"), "DEFAULT 'abc'");
        winqEqual(new ColumnConstraint().defaultTo((ExpressionConvertible)null), "DEFAULT NULL");
        winqEqual(new ColumnConstraint().collate("BINARY"), "COLLATE BINARY");
    }
}

File: src/rust/examples/tests/winq/column_constraint_test.rs
#[cfg(test)]
pub mod column_constraint_test {
    use crate::base::winq_tool::WinqTool;
    use wcdb::winq::column_constraint::ColumnConstraint;
    use wcdb::winq::conflict_action::ConflictAction;

    #[test]
    pub fn test() {
        WinqTool::winq_equal(ColumnConstraint::new(None).primary_key(), "PRIMARY KEY");
        WinqTool::winq_equal(
            ColumnConstraint::new(Some("testColumnConstraint")).primary_key(),
            "CONSTRAINT testColumnConstraint PRIMARY KEY",
        );
        WinqTool::winq_equal(
            ColumnConstraint::new(None).primary_key().auto_increment(),
            "PRIMARY KEY AUTOINCREMENT",
        );
        WinqTool::winq_equal(
            ColumnConstraint::new(Some("testColumnConstraint")).not_null(),
            "CONSTRAINT testColumnConstraint NOT NULL",
        );

        WinqTool::winq_equal(
            ColumnConstraint::new(None)
                .not_null()
                .conflict(ConflictAction::Abort),
            "NOT NULL ON CONFLICT ABORT",
        );

        WinqTool::winq_equal(
            ColumnConstraint::new(Some("testColumnConstraint")).unique(),
            "CONSTRAINT testColumnConstraint UNIQUE",
        );
        WinqTool::winq_equal(
            ColumnConstraint::new(Some("testColumnConstraint")).un_index(),
            "CONSTRAINT testColumnConstraint UNINDEXED",
        );
        WinqTool::winq_equal(ColumnConstraint::new(None).default_to(1), "DEFAULT 1");
        WinqTool::winq_equal(
            ColumnConstraint::new(None).default_to(false),
            "DEFAULT FALSE",
        );
        WinqTool::winq_equal(
            ColumnConstraint::new(None).default_to("abc"),
            "DEFAULT 'abc'",
        );
        // todo dengxudong 缺逻辑，重要，不紧急
        // WinqTool::winq_equal(ColumnConstraint::new().default_to(ExpressionConvertible), "DEFAULT NULL");
        WinqTool::winq_equal(
            ColumnConstraint::new(None).collate("BINARY"),
            "COLLATE BINARY",
        );
    }
}
